ZIP 320: Specify Alternative 2 in terms of a MUST-understand metadata item (#766)
zcash/zips: zip-0320.rst

   must be able to generate a receiving address such that only transparent
   funds will be spent in transactions with an output to this address.
2. Wallets and other Consumers that have not been upgraded to recognize the new

>    address format cannot mistake the address for another addRess type or

Reviewer: Remove this change.

Author: Huh, I have no idea how this snuck in. Fat fingers.

Motivations for Alternative 2
-----------------------------

Traceable Unified Addresses fit into the Zcash Unified Address ecosystem
defined by ZIP 316, Revision 1 [#zip-0316-revision-1]_. Existing Consumers of
Unified Addresses will not be able to send to these address unless they update
their code to understand the new MUST-understand Metadata Typecode defined in
this ZIP. 

By integrating with the Unified Address framework, it becomes possible for the
addresses being generated to include extra metadata; in particular, metadata
items such as an Address Expiry Height or Address Expiry Date
[#zip-0316-address-expiry]_ may be included. For exchange use cases such as
Binance's, it is useful to ensure that an address provided to a user has a
limited utility life, such that after expiration the user must obtain a new
address in order to be able to continue to send funds
[#binance-address-expiry]_.

Specification (Alternative 2)
-----------------------------

Upon activation of this ZIP, the section `Metadata Items` of ZIP 316
[#zip-0316-metadata-items]_ will be modified to define a new MUST-understand
Metadata Item type: Source Restriction Metadata, having Typecode

> :math:`\mathtt{0xE2}`, the value of which must be a single byte:

Reviewer: Should it always be a single byte, or should it just be constrained currently to be a single byte if that first byte is 0x00? I guess it doesn't matter which way it is specified now, as parsers need to understand both the typecode and its value, so a different-length value is no different to a different-valued single-byte value.

Author: :heavy_check_mark:

> :math:`\mathtt{0xE2}`, the value of which must be a single byte:

Reviewer 2: ```suggestion
:math:`\mathtt{0xE2}`, the value of which MUST be a single byte:
```

* :math:`\mathtt{0x00}` - Transparent Source Only

> Additional Source Restriction Metadata values may be defined in the future.

Reviewer 2: ```suggestion
Additional Source Restriction Metadata values can be defined in the future,
but a Consumer that does not recognise the value MUST reject the entire
UA as invalid.
```
In practice I think we want to define the "shielded-only" values at the same time, but that doesn't need to be in this PR.

Reviewer 2: Metadata Items in UIVKs are [not implicitly preserved](https://github.com/zcash/zips/pull/765/files#diff-8ff1060cb8fa47eb429fb6c849f1364e28fd4b42868ae8f56dd20ea8b9028b2cR856-R858) when deriving a UA; that has to be specified in the definition of each Metadata Typecode.
```suggestion
Additional Source Restriction Metadata values can be defined in the future,
but a Consumer that does not recognise the value MUST reject the entire
UA / UIVK as invalid.
```

  such that a Unified Address MUST **either**:

  * contain at least one shielded Receiver (Typecodes :math:`\mathtt{0x02}`
    and :math:`\mathtt{0x03}`), and **no** Source Restriction Metadata Item
    having value :math:`\mathtt{0x00}`; **or**
  * contain **both** a Transparent Receiver (Typecode :math:`\mathtt{0x00}` or
    Typecode :math:`\mathtt{0x01}`) **and** a Source Restriction Metadata Item
    (Typecode :math:`\mathtt{0xE2}`) having value :math:`\mathtt{0x00}`, and
    **no other** Receivers.

  A Unified Viewing Key MUST contain at least one shielded Item (Typecodes

>   :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`). 

Reviewer 2: ```suggestion
  :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`), and **no** Source
  Restriction Metadata Item.
```
This excludes even future Source Restriction Metadata Item values from being applicable to UVKs. I think that's what we want.

  such that a Unified Address MUST **either**:

  * contain at least one shielded Receiver (Typecodes :math:`\mathtt{0x02}`

>     and :math:`\mathtt{0x03}`), and **no** Source Restriction Metadata Item

Reviewer: ```suggestion
    and :math:`\mathtt{0x03}`, or any future shielded Receiver typecode), and **no** Source Restriction Metadata Item
```
or otherwise adjust the wording to not appear to require that it can only be one of those two specific receiver codes, and never any future shielded Receiver typecode.

Author: The previous wording specified exactly these two typecodes, so this is preserving that restriction; an additional change to that part of the specification is out of scope for this ZIP.

   two-byte address prefix is not :math:`[\mathtt{0x1C}, \mathtt{0xB8}]`,
   return an error. Otherwise, let the **validating key hash** be the remaining
   20 bytes of the array after removing the two-byte address prefix.
3. Construct a new Revision 1 Unified Address using a single P2PKH Receiver
   :math:`\mathtt{0x04}` with the 20-byte **validating key hash** as its value,
   and a Source Restriction Metadata Item (Typecode :math:`\mathtt{0xE2}`)
   having value :math:`\mathtt{0x00}` (Transparent Source Only). In addition,
   metadata items such as an Address Expiry Height or Address Expiry Date
   [#zip-0316-address-expiry]_ MAY be included.

> 4. Encode the Unified Address using the “``u+``” Human Readable Part as

Reviewer: Noting that `u+` is a placeholder here, until we have decided on what HRPs to use.

Author: :heavy_check_mark:

   specified for Revision 1 of ZIP 316 [#zip-0316-unified-addresses]_.

For Testnet addresses, the required lead bytes of a P2PKH address in step 2
are :math:`[\mathtt{0x1D}, \mathtt{0x25}]`.

The HRP of the resulting Unified Address is the same as for any other Revision 1
Unified Address on the relevant network as specified in [#zip-0316-revision-1]_,

> i.e. ``"u+"`` for Mainnet and ``"u+test"`` for Testnet.

Reviewer: Ditto, placeholders.

> Wallets and other Senders sending MUST ensure that only transparent UTXOs are

Reviewer: "Wallets and other Senders sending"... to what?

   20 bytes of the array after removing the two-byte address prefix.
3. Construct a new Revision 1 Unified Address using a single P2PKH Receiver
   :math:`\mathtt{0x04}` with the 20-byte **validating key hash** as its value,
   and a Source Restriction Metadata Item (Typecode :math:`\mathtt{0xE2}`)
   having value :math:`\mathtt{0x00}` (Transparent Source Only). In addition,
   metadata items such as an Address Expiry Height or Address Expiry Date
   [#zip-0316-address-expiry]_ MAY be included.
4. Encode the Unified Address using the “``u+``” Human Readable Part as
   specified for Revision 1 of ZIP 316 [#zip-0316-unified-addresses]_.

For Testnet addresses, the required lead bytes of a P2PKH address in step 2
are :math:`[\mathtt{0x1D}, \mathtt{0x25}]`.

The HRP of the resulting Unified Address is the same as for any other Revision 1
Unified Address on the relevant network as specified in [#zip-0316-revision-1]_,
i.e. ``"u+"`` for Mainnet and ``"u+test"`` for Testnet.

Wallets and other Senders MUST ensure that only transparent UTXOs are spent in
the creation of a transaction to any Unified Address containing a Source
Restriction Metadata Item having value :math:`\mathtt{0x00}`.

> 

Reviewer 2: ```suggestion

Any Source Restriction Metadata Item MUST be preserved with the same value
when deriving a UIVK from a UFVK, or a UA from a UIVK. It has no other effect
on the meaning of the UFVK or UIVK.

```
